### PR TITLE
Add train() / eval() / is_training() to C++ ScriptModule API

### PR DESCRIPTION
**Pull request by yf225**
This PR aims to fix https://discuss.pytorch.org/t/how-to-change-a-loaded-model-to-evaluation-mode-in-c/32330, by adding `train()` / `eval()` / `is_training()` to C++ ScriptModule API.

### CODE REVIEW

File: test/cpp/jit/tests.h [hunk 1]
```diff
@@ -2067,6 +2068,17 @@ void testAliasAnalysis() {
         usesB->node(), mutatesAliasOfB->node()));
   }
 }
+
+void testEvalModeForLoadedModule() {
+  std::string module_path = "dropout_model.pt";
+  std::shared_ptr<torch::jit::script::Module> module = torch::jit::load(module_path);
```
Comment (goldsborough) [inline test/cpp/jit/tests.h]: Maybe first assert that we start out in training mode `JIT_ASSERT(module->get_module("dropout")->is_training());`

File: test/cpp/jit/tests.h [hunk 2]
```diff
@@ -2067,6 +2068,17 @@ void testAliasAnalysis() {
...
+  // Test eval mode
```
Comment (goldsborough) [inline test/cpp/jit/tests.h]: I think the comments here don't add too much

File: test/cpp/jit/tests_setup.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import torch
+
+
+def testEvalModeForLoadedModule_setup():
```
Comment (goldsborough) [inline test/cpp/jit/tests_setup.py]: Is there a reason this file isn't just a straight-line script? I.e. without the function, `main` and `if __name__ ...`, just the contents of `testEvalModeForLoadedModule_setup`

File: torch/csrc/jit/script/module.h [hunk 4]
```diff
@@ -481,6 +481,25 @@ struct Module {
     }
     fn(*this);
   }
+  void train(bool on = true) {
+    for (auto& submod : get_modules()) {
+      submod.value().module->train(on);
```
Comment (goldsborough) [inline torch/csrc/jit/script/module.h]: you can also write `submod->module->train(on)`

File: torch/csrc/jit/script/module.h [hunk 5]
```diff
@@ -481,6 +481,25 @@ struct Module {
...
+    }
+    auto t = autograd::make_variable(at::full({}, on ? 1 : 0, at::kLong));
```
Comment (goldsborough) [inline torch/csrc/jit/script/module.h]: You can replace
```
auto t = autograd::make_variable(at::full({}, on ? 1 : 0, at::kLong));
```
with
```
auto t = torch::tensor(on ? 1 : 0, at::kLong);
```

To clarify: `torch::` factory functions create variables while `at::` functions create tensors. `torch::tensor`/`at::tensor` is like `torch.tensor` in python (creates a tensor with the values you give it)`

Comment (goldsborough) [reply torch/csrc/jit/script/module.h]: And then just embed it in the `register_parameter` call:
```
register_parameter("training", torch::tensor(on ? 1 : 0, at::kLong), /*is_buffer=*/true);
```

File: torch/csrc/jit/script/module.h [hunk 6]
```diff
@@ -481,6 +481,25 @@ struct Module {
...
+    register_parameter("training", std::move(t), true);
+  }
+  void eval() {
+    train(/*on=*/false);
+  }
+  bool is_training() {
+    if (auto p = find_parameter("training")) {
+      return (*p->slot()).item().toLong() == 1;
```
Comment (goldsborough) [inline torch/csrc/jit/script/module.h]: Here you are converting the tensor first to a `Scalar`, and then to a long. You can just use `item<T>` to get the value of the tensor without the intermediate `Scalar`:
```
(*p->slot()).item<int64_t>()
```

File: test/cpp/jit/tests.h (hunk 2 above)
Comment (yf225) [reply test/cpp/jit/tests.h]: removed

File: test/cpp/jit/tests_setup.py (hunk 3 above)
Comment (yf225) [reply test/cpp/jit/tests_setup.py]: `test_cpp_cuda` in test_jit.py also runs `runJITCPPTests()`, and we need to be able to call the test setup code in `test_cpp_cuda` to prepare the `dropout_model.pt` file needed, which is why we need script to be in a Python function.

File: torch/csrc/jit/script/module.h [hunk 7]
```diff
@@ -481,6 +482,23 @@ struct Module {
...
+      submod->module->train(on);
+    }
+    register_parameter("training", torch::tensor(on ? 1 : 0, at::kLong), /*is_buffer=*/true);
+  }
+  void eval() {
+    train(/*on=*/false);
+  }
+  bool is_training() {
+    if (auto p = find_parameter("training")) {
+      return (*p->slot()).item<int64_t>() == 1;
```
Comment (goldsborough) [inline torch/csrc/jit/script/module.h]: nit: `p->slot()->`

File: torch/csrc/jit/script/module.h [hunk 8]
```diff
@@ -481,6 +482,23 @@ struct Module {
...
+    } else {
```
Comment (goldsborough) [inline torch/csrc/jit/script/module.h]: nit: remove else block, just `return true` + comment (since you return in the `if` block)

File: torch/csrc/jit/script/module.h [hunk 9]
```diff
@@ -481,6 +482,23 @@ struct Module {
     }
     fn(*this);
   }
+  void train(bool on = true) {
```
Comment (goldsborough) [inline torch/csrc/jit/script/module.h]: nit: add a comment above. This ScriptModule is not documented unfortunately (most JIT code isn't; I've tried to tell people in the  past...). This class is user-facing and has public documentation: https://pytorch.org/cppdocs/api/structtorch_1_1jit_1_1script_1_1_module.html#exhale-struct-structtorch-1-1jit-1-1script-1-1-module

Compare with docs of `nn::Module`: https://pytorch.org/cppdocs/api/classtorch_1_1nn_1_1_module.html#exhale-class-classtorch-1-1nn-1-1-module

**Review by goldsborough — APPROVED** [review]
Looks good! Some friendly nits which you can address if you feel like it, or not
